vertexyTests/src/private/Maze.cpp

```diff
@@ -321,18 +423,57 @@ int MazeSolver::solve(int times, int numRows, int numCols, int seed, bool printV
 
 		auto selfStepTile = make_shared<TVertexToDataGraphRelation<VarID>>(stepData);
 
-		// If this tile is the entrance in the maze, constrain it to be the entrance in this step.
+		auto stepDoorData = solver.makeVariableGraph(stepName, ITopology::adapt(grid), stepDomain, { wstring::CtorSprintf(), TEXT("StepDoor%d-"), step });
+		stepDoorDatas.push_back(stepDoorData);
+		auto selfStepDoorTile = make_shared<TVertexToDataGraphRelation<VarID>>(stepDoorData);
+		// constraint 
 		solver.makeGraphConstraint<IffConstraint>(grid,
-			GraphRelationClause(selfStepTile, step_Origin),
-			vector{GraphRelationClause(selfTile, cell_Entrance)}
+			GraphRelationClause(selfStepDoorTile, step_Origin),
+			vector{ GraphRelationClause(selfTile, step == 0 ? cell_Entrance : vector<int>{cell_Doors[step - 1]})}
+		);
+
+		// if stepDoor != reachable, tile can't be a door
+		solver.makeGraphConstraint<ClauseConstraint>(grid, ENoGood::NoGood,
+					GraphRelationClause(selfStepDoorTile, EClauseSign::Outside, step_Reachable),
+					GraphRelationClause(selfTile, step == NUM_KEYS ? cell_Exit : vector{ cell_Keys[step] })
 		);
 
+		if (TEST_SHORTEST_PATH)
+		{
+			//if this is the most recent door, constrain it to the be the entrance in this step
+			//solver.makeGraphConstraint<IffConstraint>(grid,
+			//	GraphRelationClause(selfStepTile, step_door),
+			//	vector{ GraphRelationClause(selfTile, step == 0 ? cell_Entrance : vector<int>{cell_Doors[step - 1]})}
+			//);
+
+			solver.makeGraphConstraint<IffConstraint>(grid,
+					GraphRelationClause(selfStepTile, step_Origin),
+						vector{ GraphRelationClause(selfTile, cell_Entrance) }
+			);
+		}
+		else
+		{
+			//If this tile is the entrance in the maze, constrain it to be the entrance in this step.
+			solver.makeGraphConstraint<IffConstraint>(grid,
+					GraphRelationClause(selfStepTile, step_Origin),
+						vector{ GraphRelationClause(selfTile, cell_Entrance) }
+			);
+		}
+
 		// A tile can never be passable in any step if it is a wall
 		solver.makeGraphConstraint<ClauseConstraint>(grid, ENoGood::NoGood,
 			GraphRelationClause(selfStepTile, step_ReachableOrOrigin),
 			GraphRelationClause(selfTile, cell_Wall)
 		);
 
+		if (TEST_SHORTEST_PATH)
+		{
+			solver.makeGraphConstraint<ClauseConstraint>(grid, ENoGood::NoGood,
+				GraphRelationClause(selfStepDoorTile, step_ReachableOrOrigin),
+				GraphRelationClause(selfTile, cell_Wall)
+			);
+		}
+
 		// If we don't have all the keys at this step...
 		if (step < NUM_KEYS)
 		{

```

Comment:
I believe the constraint after this line is redundant if TEST_SHORTEST_PATH is on